salt/utils/user.py

```diff
@@ -295,9 +295,7 @@ def get_group_list(user, include_default=True):
         try:
             user_group_list = os.getgrouplist(user, pwd.getpwnam(user).pw_gid)
             group_names = [
```

Comment:
**suggestion (blocking):** if we're talking about performance considerations, it looks like moving `grp.getgrall()` out of the list comprehension should be the first step. have you tried that?

```suggestion
            grall = grp.getgrall()
            group_names = [
                _group.gr_name
                for _group in grall
                if _group.gr_gid in user_group_list
              ]
```

Reply:
**question (blocking):** Ah. I'm reading #64953 now. So, is the alternate flow proposed here intended to correct the remote group omission in that issue?

Reply:
This is not my PR so I can't comment to the original intent ... but the proposal here effectively reverts the code back to its state prior to #62378.  I think:

- the writers intent may be that on systems with remote groups or `sssd` with `enumeration` **on**, then `grp.grall` can take a very long time (#64888); and
- with `enumeration` **off**, it won't take a long time but will only return local groups (which is the issue I have ... #64953)

(Its not clear to me whether it is in fact possible to deal with the possibility of duplicate ids without using `grp.grall` ... but my bias is obviously that its more important to get all of a users groups rather than handle the edge case of duplicate `gid`)

Reply:
Bias noted.  :smile:    I'd like to see if we can still solve for both.

Reply:
at the end of the day if we can not handle both. the duplicate gid one is the lease common use and one not fully supported by the tools within linux. [commands like `groups` or `who` also don't work with displaying duplicate gids, and `groupadd` will not allow adding a duplicate gid]. 

the only way i see to handle both use cases is not pretty, or performant. it would take grabbing from both pieces of information. grabbing the `grall` parsing that down to the user then using the looped `greint` to fetch anything not grabbed by the `grall`. and finally dedupping the compiled list

Reply:
found a way to retain duplicate GID checks for local groups, which is really all we can expect. @wcbzero , can you integrate this patch, or do you want to close the PR and I'll open another?

```diff
diff --git a/salt/utils/user.py b/salt/utils/user.py
index 9dc1dfd2ef..1ee6323445 100644
--- a/salt/utils/user.py
+++ b/salt/utils/user.py
@@ -10,6 +10,7 @@ import logging
 import os
 import sys
 
+import salt.utils.files
 import salt.utils.path
 import salt.utils.platform
 import salt.utils.stringutils
@@ -293,10 +294,29 @@ def get_group_list(user, include_default=True):
         # Try os.getgrouplist, available in python >= 3.3
         log.trace("Trying os.getgrouplist for '%s'", user)
         try:
-            user_group_list = os.getgrouplist(user, pwd.getpwnam(user).pw_gid)
-            group_names = [
-                grp.getgrgid(group_id).gr_name for group_id in user_group_list
+            user_group_list = sorted(os.getgrouplist(user, pwd.getpwnam(user).pw_gid))
+            local_grall = _getgrall()
+            local_gids = sorted(lgrp.gr_gid for lgrp in local_grall)
+            max_idx = -1
+            local_max = local_gids[max_idx]
+            while local_max >= 65000:
+                max_idx -= 1
+                local_max = local_gids[max_idx]
+            user_group_list_local = [
+                lgrp for lgrp in user_group_list if lgrp <= local_max
             ]
+            user_group_list_remote = [
+                rgrp for rgrp in user_group_list if rgrp > local_max
+            ]
+            local_group_names = [
+                _group.gr_name
+                for _group in local_grall
+                if _group.gr_gid in user_group_list_local
+            ]
+            remote_group_names = [
+                grp.getgrgid(group_id).gr_name for group_id in user_group_list_remote
+            ]
+            group_names = local_group_names + remote_group_names
         except Exception:  # pylint: disable=broad-except
             pass
     elif HAS_PYSSS:
@@ -383,3 +403,24 @@ def get_gid(group=None):
             return grp.getgrnam(group).gr_gid
         except KeyError:
             return None
+
+
+def _getgrall(root=None):
+    """
+    Alternative implemetantion for getgrall, that use only /etc/group
+    """
+    ret = []
+    root = "/" if not root else root
+    etc_group = os.path.join(root, "etc/group")
+    with salt.utils.files.fopen(etc_group) as fp_:
+        for line in fp_:
+            line = salt.utils.stringutils.to_unicode(line)
+            comps = line.strip().split(":")
+            # Generate a getgrall compatible output
+            comps[2] = int(comps[2])
+            if comps[3]:
+                comps[3] = [mem.strip() for mem in comps[3].split(",")]
+            else:
+                comps[3] = []
+            ret.append(grp.struct_group(comps))
+    return ret
```